clusterloader/docs/vision.md

```diff
@@ -0,0 +1,334 @@
+
+We would like to get back to that and build fully functional Cluster Loader and
+use it as a framework to run all our scalability tests. This doc is describing
+high-level vision of how this will work.
+
+
+## Vision
+
+At the high level, a single test will consist of a number of steps. If we have
+really good usecase, we may consider creating a DAG of steps, but for now we
+don't have any rationale for it. In each of those steps we will be creating,
+updating and/or deleting a number of different objects.
+Additionally, we will introduce a set of predefined operations (that user will be
+able to use as phases). They will allow users to monitor/measure performance impact
+of user-defined phases.
+The following subsections will describe this in a bit more detail.
+
+### Config
+
+A single test scenario will be defined by a `Config`. Its schema will be as following:
+
+```
+struct Config {
+	// TBD: Some config allowing to access cluster, e.g. path to kubeconfig file.
+
+	// Number of namespace in the test.
+	Namespaces int32
+	// Steps of the test.
+	Steps []Step
+}
+```
+
+With a test being defined by a single json/yaml file, it should be pretty simple
+to modify scenarios, and fork them to new ones.
+
+Note that before running any steps, ClusterLoader will create all the necessary
+namespaces and after running all of them will delete them (together with all
+objects that remained undeleted after running the test). Namespaces are described
+in more details in the later part of this document.
+
+### Step
+
+Each step will consist of a number of create, update and delete operations (potentially
+on many different object types) or a number of monitoring/measurement-related actions.
+A single step is defined as following:
+
+```
+struct Step {
+	// Only one of these can be non-empty.
+	Phases       []Phase
+	Measurements []string
+}
+```
+
+We make `Phases` and `Measurements` separate concepts to ensure correct ordering
+between those two types of actions. It's very important to ensure that proper
+measurements are started before we start given actions and they are stopped when
+all actions are done.
+
+Also note that all `Phases` and `Measurements` within a single `Step` will be
+run in parallel - a `Step` ends when all its `Phases` or `Measurements` finish.
+
+### Phase
+
+Phase declaratively defines state of objects we should reach in the underlying
+Kubernetes cluster. A single declaration may result in a number of create, update
+and delete operations depending on the current state of the cluster.
+
+We define the phase as following:
+
+```
+struct Phase {
+	// Set of namespaces in which objects should be reconciled.
+	// If null, objects are assumed to be cluster scoped.
+	NamespaceRange *NamespaceRange
+	// Number of instances of a given object to exist in each
+	// of referenced namespaces.
+	ReplicasPerNamespace int32
+	// Name of TuningSet to be used.
+	TuningSet string
+	// A set of objects that should be reconciled.
+	Objects []Object
```

Comment:
I'm not sure I understand why this should be an array?
This would iiuc mean that we use same `ReplicasPerNamespace` to be applied for all the objects in the array (which I can't see much use of).

Reply:
Also same applies wrt other configs like TuningSet and NamespaceRange.

Reply:
Ok... I see you have an explanation below. Let me read it.

Reply:
Read it and I still have some confusion. So IIUC:

- we already have steps that are serial
- within a step we have phases that are parallel
- with a phase we have objects that are again serially processed

What I don't understand is why it isn't enough to have just one serial construct and one parallel construct? Maybe we can discuss this f2f.

Reply:
Discussed f2f and added rationale below.